lib/wallaroo/core/source/connector_source/connector_framed_source_notify.pony

```diff
@@ -82,91 +254,348 @@ class ConnectorSourceNotify[In: Any val]
+                // _handler.decode(bytes) failed
+                if m.message is None then
+                  // TODO [post-source-migration] revisit this error message
+                  return _to_error_state(source, "No message bytes and BOUNDARY not set")
+                end
+                @printf[I32](("Unable to decode message at " + _pipeline_name + " source\n").cstring())
+                ifdef debug then
+                  Fail()
+                end
+                return _to_error_state(source, "Unable to decode message")
+              end
+            end
+          else
+            return _to_error_state(source, "Unknown StreamId")
           end
-          (true, ingest_ts)
         end
 
-      if is_finished then
-        let end_ts = WallClock.nanoseconds()
-        let time_spent = end_ts - ingest_ts
+      | let m: cwm.AckMsg =>
+        ifdef "trace" then
+          @printf[I32]("TRACE: got AckMsg\n".cstring())
+        end
+        return _to_error_state(source, "Invalid message: ack")
 
-        ifdef "detailed-metrics" then
-          _metrics_reporter.step_metric(_pipeline_name,
-            "Before end at Connector Source", 9999,
-            last_ts, end_ts)
+      | let m: cwm.RestartMsg =>
+        ifdef "trace" then
+          @printf[I32]("TRACE: got RestartMsg\n".cstring())
         end
+        return _to_error_state(source, "Invalid message: restart")
 
-        _metrics_reporter.pipeline_metric(_pipeline_name, time_spent +
-          pipeline_time_spent)
-        _metrics_reporter.worker_metric(_pipeline_name, time_spent)
+      end
+    else
+      @printf[I32](("Unable to decode message at " + _pipeline_name +
+        " source\n").cstring())
+      ifdef debug then
+        Fail()
+      end
+      return _to_error_state(source, "Unable to decode message")
+    end
+    _continue_perhaps(source)
+
+  fun ref _run_and_subsequent_activity(latest_metrics_id: U16,
+    ingest_ts: U64,
+    pipeline_time_spent: U64,
+    key_string: String val,
+    source: ConnectorSource[In] ref,
+    decoded: (In val| None val),
+    s: _StreamState,
+    message_id: (cwm.MessageId|None),
+    flags: cwm.Flags): Bool
+   =>
+    let decode_end_ts = WallClock.nanoseconds()
+    _metrics_reporter.step_metric(_pipeline_name,
+      "Decode Time in Connector Source", latest_metrics_id, ingest_ts,
+      decode_end_ts)
+    let latest_metrics_id' = latest_metrics_id + 1
+
+    let msg_uid = _msg_id_gen()
+
+    // TODO: We need a way to determine the key based on the policy
+    // for any particular connector. For example, the Kafka connector
+    // needs a way to provide the Kafka key here.
+
+    let initial_key =
+      if key_string isnt None then
+        key_string
+      else
+        // wat
+        // TODO [post-source-migration] should this be stream_id for messages that
+        // do not provide a key?
+        // What is a sane default here?
+        // What does it mean to use a unique key for each message?
```

Comment:
It means that if the next stage is stateless, they'll be effectively distributed at random among the available workers.  If the next stage is stateful, they'll each be routed to their own state, meaning the pipeline definition will need to precede the stateful stage by either `key_by` or `collect` to get reasonable behavior.